### PR TITLE
Statically eliminate redundant sygus constructors

### PR DESCRIPTION
**Pull request by ajreynol**
Moves most of this code to grammar normalization out of Sygus term database, work towards #1235.

### CODE REVIEW

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 1]
```diff
@@ -516,6 +516,18 @@ Node DatatypesRewriter::mkTester(Node n, int i, const Datatype& dt)
...
+Node DatatypesRewriter::mkSplit(Node n, const Datatype& dt)
+{
+  std::vector<Node> splits;
+  for (unsigned i = 0, ncons = dt.getNumConstructors(); i < ncons; i++)
+  {
+    Node test = mkTester(n, i, dt);
+    splits.push_back(test);
+  }
+  NodeManager* nm = NodeManager::currentNM();
+  return splits.size() == 1 ? splits[0] : nm->mkNode(kind::OR, splits);
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: seems spurious to not use ``NodeManager::currentNM()`` directly here.

File: src/theory/quantifiers/sygus_grammar_red.h [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+   * this constructor.
+   * 2 : redundant since there exists a term in the enumerator of the registered
+   * type that is equivalent to terms of this constructor.
+   *
+   * For example, for grammar:
+   *   A -> C > B | B < C | not D
+   *   B -> x | y
+   *   C -> 0 | 1 | C+C
+   *   D -> B >= C
+   * If A is register with this class, then we store may store { 0, 1, 2 },
```
Comment (HanielB) [inline src/theory/quantifiers/sygus_grammar_red.h]: register --> registered

File: src/theory/quantifiers/sygus_grammar_red.h [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  std::map<Node, unsigned> d_gen_cons;
+  /** get generic list
+   *
+   * This function constructs all well-typed variants of a term of the form
+   *    op( x1, ..., xn )
+   * where op is the builtin operator for dt[c], and xi = pre[i] for i=1,...,n.
+   *
+   * It constructs a list of terms of the form g * sigma, where sigma
+   * is an automorphism on { x1...xn } such that for all xi -> xj in sigma,
+   * the sygus for arguments i and j of dt[c] are the same. We store this
```
Comment (HanielB) [inline src/theory/quantifiers/sygus_grammar_red.h]: the sygus ...?

File: src/theory/datatypes/datatypes_rewriter.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/datatypes/datatypes_rewriter.cpp]: This is to make it fit to 80 lines, which looks cleaner to me.

Comment (HanielB) [reply src/theory/datatypes/datatypes_rewriter.cpp]: I see

**Pull request merged**